python/ray/tests/test_task_events_2.py

```diff
@@ -289,160 +301,236 @@ def verify():
+@pytest.mark.parametrize(
+    "exit_type",
+    ["exit_kill", "exit_exception"],
+)
+@pytest.mark.parametrize(
+    "actor_or_normal_tasks",
+    ["normal_task", "actor_task"],
+)
+def test_fault_tolerance_chained_task_fail(
+    shutdown_only, exit_type, actor_or_normal_tasks
+):
+
+    ray.init(_system_config=_SYSTEM_CONFIG)
+
+    def sleep_or_fail(pid_actor=None, exit_type=None):
+        if exit_type is None:
+            time.sleep(999)
+        # Wait until the children run
+        if pid_actor:
+            wait_for_condition(
+                lambda: len(ray.get(pid_actor.get_pids.remote())) == 3,
+            )
+
+        if exit_type == "exit_kill":
+            os._exit(1)
+        else:
+            raise ValueError("Expected to fail")
+
+    # Test a chain of tasks
+    @ray.remote(max_retries=0)
+    def A(exit_type, pid_actor):
+        x = B.remote(pid_actor)
+        ray.get(pid_actor.report_pid.remote("A", os.getpid()))
+        sleep_or_fail(pid_actor, exit_type)
+        ray.get(x)
+
+    @ray.remote(max_retries=0)
+    def B(pid_actor):
+        x = C.remote(pid_actor)
+        ray.get(pid_actor.report_pid.remote("B", os.getpid()))
+        sleep_or_fail()
+        ray.get(x)
+
+    @ray.remote(max_retries=0)
+    def C(pid_actor):
+        ray.get(pid_actor.report_pid.remote("C", os.getpid()))
+        sleep_or_fail()
+
+    @ray.remote(max_restarts=0, max_task_retries=0)
+    class Actor:
+        def run(self, pid_actor):
+            with pytest.raises(
+                (ray.exceptions.RayTaskError, ray.exceptions.WorkerCrashedError)
+            ):
+                ray.get(A.remote(exit_type=exit_type, pid_actor=pid_actor))
+
+    pid_actor = PidActor.remote()
+
+    def verify():
+        task_names_to_pid = ray.get(pid_actor.get_pids.remote())
+        assert len(task_names_to_pid) == 3, task_names_to_pid
+        return verify_tasks_running_or_terminated(task_names_to_pid)
+
+    if actor_or_normal_tasks == "normal_task":
+        with pytest.raises(
+            (ray.exceptions.RayTaskError, ray.exceptions.WorkerCrashedError)
+        ):
+            ray.get(A.remote(exit_type=exit_type, pid_actor=pid_actor))
+    else:
+        a = Actor.remote()
+        ray.get(a.run.remote(pid_actor=pid_actor))
 
-    # Some constants
-    NORMAL_TASK = 0
-    ACTOR_TASK = 1
+    wait_for_condition(verify)
 
+
+@pytest.mark.parametrize(
+    "death_list",
+    [
+        [("A", "exit_kill")],
+        [("Abb", "exit_kill"), ("C", "exit_exception")],
+        [("D", "exit_kill"), ("Ca", "exit_kill"), ("A", "exit_exception")],
+    ],
+)
+def test_fault_tolerance_advanced_tree(shutdown_only, death_list):
+    """
+    Test fault tolerance for a more complicated execution graph.
+    """
     # Root should always be finish
     execution_graph = {
         "root": [
-            (NORMAL_TASK, "A"),
-            (ACTOR_TASK, "B"),
-            (NORMAL_TASK, "C"),
-            (ACTOR_TASK, "D"),
+            ("normal_task", "A"),
```

Comment:
Using string such that it shows up better in the test output, instead of 0/1, it's now "normal_task"/"actor_task".